File: jscomp/stdlib-406/stream.ml
(**************************************************************************)
(*                                                                        *)
(*                                 OCaml                                  *)
(*                                                                        *)
(*         Daniel de Rauglaudre, projet Cristal, INRIA Rocquencourt       *)
(*                                                                        *)
(*   Copyright 1997 Institut National de Recherche en Informatique et     *)
(*     en Automatique.                                                    *)
(*                                                                        *)
(*   All rights reserved.  This file is distributed under the terms of    *)
(*   the GNU Lesser General Public License version 2.1, with the          *)
(*   special exception on linking described in the file LICENSE.          *)
(*                                                                        *)
(**************************************************************************)

type 'a t = 'a cell option
and 'a cell = { mutable count : int; mutable data : 'a data }
and 'a data =
    Sempty
  | Scons of 'a * 'a data
  | Sapp of 'a data * 'a data
  | Slazy of 'a data Lazy.t
  | Sgen of 'a gen
  | Sbuffio : buffio -> char data
and 'a gen = { mutable curr : 'a option option; func : int -> 'a option }
and buffio =
  { ic : in_channel; buff : bytes; mutable len : int; mutable ind : int }

exception Failure
exception Error of string

let count = function
  | None -> 0
  | Some { count } -> count
let data = function
  | None -> Sempty
  | Some { data } -> data

let fill_buff b =
  b.len <- input b.ic b.buff 0 (Bytes.length b.buff); b.ind <- 0


let rec get_data : type v. int -> v data -> v data = fun count d -> match d with
 (* Returns either Sempty or Scons(a, _) even when d is a generator
    or a buffer. In those cases, the item a is seen as extracted from
 the generator/buffer.
 The count parameter is used for calling `Sgen-functions'.  *)
   Sempty | Scons (_, _) -> d
 | Sapp (d1, d2) ->
     begin match get_data count d1 with
       Scons (a, d11) -> Scons (a, Sapp (d11, d2))
     | Sempty -> get_data count d2
     | _ -> assert false
     end
 | Sgen {curr = Some None} -> Sempty
 | Sgen ({curr = Some(Some a)} as g) ->
     g.curr <- None; Scons(a, d)
 | Sgen g ->
     begin match g.func count with
       None -> g.curr <- Some(None); Sempty
     | Some a -> Scons(a, d)
         (* Warning: anyone using g thinks that an item has been read *)
     end
 | Sbuffio b ->
     if b.ind >= b.len then fill_buff b;
     if b.len == 0 then Sempty else
       let r = Bytes.unsafe_get b.buff b.ind in
       (* Warning: anyone using g thinks that an item has been read *)
       b.ind <- succ b.ind; Scons(r, d)
 | Slazy f -> get_data count (Lazy.force f)


let rec peek_data : type v. v cell -> v option = fun s ->
 (* consult the first item of s *)
 match s.data with
   Sempty -> None
 | Scons (a, _) -> Some a
 | Sapp (_, _) ->
     begin match get_data s.count s.data with
       Scons(a, _) as d -> s.data <- d; Some a
     | Sempty -> None
     | _ -> assert false
     end
 | Slazy f -> s.data <- (Lazy.force f); peek_data s
 | Sgen {curr = Some a} -> a
 | Sgen g -> let x = g.func s.count in g.curr <- Some x; x
 | Sbuffio b ->
     if b.ind >= b.len then fill_buff b;
     if b.len == 0 then begin s.data <- Sempty; None end
     else Some (Bytes.unsafe_get b.buff b.ind)


let peek = function
  | None -> None
  | Some s -> peek_data s


let rec junk_data : type v. v cell -> unit = fun s ->
  match s.data with
    Scons (_, d) -> s.count <- (succ s.count); s.data <- d
  | Sgen ({curr = Some _} as g) -> s.count <- (succ s.count); g.curr <- None
  | Sbuffio b -> s.count <- (succ s.count); b.ind <- succ b.ind
  | _ ->
      match peek_data s with
        None -> ()
      | Some _ -> junk_data s


let junk = function
  | None -> ()
  | Some data -> junk_data data

let rec nget_data n s =
  if n <= 0 then [], s.data, 0
  else
    match peek_data s with
      Some a ->
        junk_data s;
        let (al, d, k) = nget_data (pred n) s in a :: al, Scons (a, d), succ k
    | None -> [], s.data, 0


let npeek_data n s =
  let (al, d, len) = nget_data n s in
  s.count <- (s.count - len);
  s.data <- d;
  al


let npeek n = function
  | None -> []
  | Some d -> npeek_data n d

let next s =
  match peek s with
    Some a -> junk s; a
  | None -> raise Failure


let empty s =
  match peek s with
    Some _ -> raise Failure
  | None -> ()


let iter f strm =
  let rec do_rec () =
    match peek strm with
      Some a -> junk strm; ignore(f a); do_rec ()
    | None -> ()
  in
  do_rec ()


(* Stream building functions *)

let from f = Some {count = 0; data = Sgen {curr = None; func = f}}

let of_list l =
  Some {count = 0; data = List.fold_right (fun x l -> Scons (x, l)) l Sempty}


let of_string s =
  let count = ref 0 in
  from (fun _ ->
    (* We cannot use the index passed by the [from] function directly
       because it returns the current stream count, with absolutely no
       guarantee that it will start from 0. For example, in the case
       of [Stream.icons 'c' (Stream.from_string "ab")], the first
       access to the string will be made with count [1] already.
    *)
    let c = !count in
    if c < String.length s
    then (incr count; Some s.[c])
    else None)


let of_bytes s =
  let count = ref 0 in
  from (fun _ ->
    let c = !count in
    if c < Bytes.length s
    then (incr count; Some (Bytes.get s c))
    else None)


let of_channel ic =
  Some {count = 0;
        data = Sbuffio {ic = ic; buff = Bytes.create 4096; len = 0; ind = 0}}


(* Stream expressions builders *)

let iapp i s = Some {count = 0; data = Sapp (data i, data s)}
let icons i s = Some {count = 0; data = Scons (i, data s)}
let ising i = Some {count = 0; data = Scons (i, Sempty)}

let lapp f s =
  Some {count = 0; data = Slazy (lazy(Sapp (data (f ()), data s)))}

let lcons f s = Some {count = 0; data = Slazy (lazy(Scons (f (), data s)))}
let lsing f = Some {count = 0; data = Slazy (lazy(Scons (f (), Sempty)))}

let sempty = None
let slazy f = Some {count = 0; data = Slazy (lazy(data (f ())))}

(* For debugging use *)

let rec dump : type v. (v -> unit) -> v t -> unit = fun f s ->
  print_string "{count = ";
  print_int (count s);
  print_string "; data = ";
  dump_data f (data s);
  print_string "}";
  print_newline ()
and dump_data : type v. (v -> unit) -> v data -> unit = fun f ->
  function
    Sempty -> print_string "Sempty"
  | Scons (a, d) ->
      print_string "Scons (";
      f a;
      print_string ", ";
      dump_data f d;
      print_string ")"
  | Sapp (d1, d2) ->
      print_string "Sapp (";
      dump_data f d1;
      print_string ", ";
      dump_data f d2;
      print_string ")"
  | Slazy _ -> print_string "Slazy"
  | Sgen _ -> print_string "Sgen"
  | Sbuffio _ -> print_string "Sbuffio"

File: lib/es6/stream.js


import * as List from "./list.js";
import * as Caml_bytes from "./caml_bytes.js";
import * as Caml_option from "./caml_option.js";
import * as Caml_string from "./caml_string.js";
import * as Caml_exceptions from "./caml_exceptions.js";
import * as CamlinternalLazy from "./camlinternalLazy.js";

let Failure = /* @__PURE__ */Caml_exceptions.create("Stream.Failure");

let $$Error = /* @__PURE__ */Caml_exceptions.create("Stream.Error");

function count(param) {
  if (param !== undefined) {
    return param.count;
  } else {
    return 0;
  }
}

function data(param) {
  if (param !== undefined) {
    return param.data;
  } else {
    return "Sempty";
  }
}

function get_data(count, _d) {
  while (true) {
    let d = _d;
    if (typeof d !== "object") {
      return d;
    }
    switch (d.TAG) {
      case "Scons" :
        return d;
      case "Sapp" :
        let d2 = d._1;
        let match = get_data(count, d._0);
        if (typeof match !== "object") {
          _d = d2;
          continue;
        }
        if (match.TAG === "Scons") {
          return {
            TAG: "Scons",
            _0: match._0,
            _1: {
              TAG: "Sapp",
              _0: match._1,
              _1: d2
            }
          };
        }
        throw new Error("Assert_failure", {
          cause: {
            RE_EXN_ID: "Assert_failure",
            _1: [
              "stream.res",
              53,
              13
            ]
          }
        });
      case "Slazy" :
        _d = CamlinternalLazy.force(d._0);
        continue;
      case "Sgen" :
        let g = d._0;
        let match$1 = g.curr;
        if (match$1 !== undefined) {
          let a = Caml_option.valFromOption(match$1);
          if (a !== undefined) {
            g.curr = undefined;
            return {
              TAG: "Scons",
              _0: Caml_option.valFromOption(a),
              _1: d
            };
          } else {
            return "Sempty";
          }
        }
        let a$1 = g.func(count);
        if (a$1 !== undefined) {
          return {
            TAG: "Scons",
            _0: Caml_option.valFromOption(a$1),
            _1: d
          };
        } else {
          g.curr = Caml_option.some(undefined);
          return "Sempty";
        }
    }
  };
}

function peek_data(s) {
  while (true) {
    let f = s.data;
    if (typeof f !== "object") {
      return;
    }
    switch (f.TAG) {
      case "Scons" :
        return Caml_option.some(f._0);
      case "Sapp" :
        let d = get_data(s.count, s.data);
        if (typeof d !== "object") {
          return;
        }
        if (d.TAG === "Scons") {
          s.data = d;
          return Caml_option.some(d._0);
        }
        throw new Error("Assert_failure", {
          cause: {
            RE_EXN_ID: "Assert_failure",
            _1: [
              "stream.res",
              83,
              13
            ]
          }
        });
      case "Slazy" :
        s.data = CamlinternalLazy.force(f._0);
        continue;
      case "Sgen" :
        let g = f._0;
        let a = g.curr;
        if (a !== undefined) {
          return Caml_option.valFromOption(a);
        }
        let x = g.func(s.count);
        g.curr = Caml_option.some(x);
        return x;
    }
  };
}

function peek(param) {
  if (param !== undefined) {
    return peek_data(param);
  }
  
}

function junk_data(s) {
  while (true) {
    let g = s.data;
    if (typeof g === "object") {
      switch (g.TAG) {
        case "Scons" :
          s.count = s.count + 1 | 0;
          s.data = g._1;
          return;
        case "Sgen" :
          let g$1 = g._0;
          let match = g$1.curr;
          if (match !== undefined) {
            s.count = s.count + 1 | 0;
            g$1.curr = undefined;
            return;
          }
          break;
      }
    }
    let match$1 = peek_data(s);
    if (match$1 === undefined) {
      return;
    }
    continue;
  };
}

function junk(param) {
  if (param !== undefined) {
    return junk_data(param);
  }
  
}

function nget_data(n, s) {
  if (n <= 0) {
    return [
      /* [] */0,
      s.data,
      0
    ];
  }
  let a = peek_data(s);
  if (a === undefined) {
    return [
      /* [] */0,
      s.data,
      0
    ];
  }
  let a$1 = Caml_option.valFromOption(a);
  junk_data(s);
  let match = nget_data(n - 1 | 0, s);
  return [
    {
      hd: a$1,
      tl: match[0]
    },
    {
      TAG: "Scons",
      _0: a$1,
      _1: match[1]
    },
    match[2] + 1 | 0
  ];
}

function npeek(n, param) {
  if (param !== undefined) {
    let match = nget_data(n, param);
    param.count = param.count - match[2] | 0;
    param.data = match[1];
    return match[0];
  } else {
    return /* [] */0;
  }
}

function next(s) {
  let a = peek(s);
  if (a !== undefined) {
    junk(s);
    return Caml_option.valFromOption(a);
  }
  throw new Error(Failure, {
    cause: {
      RE_EXN_ID: Failure
    }
  });
}

function empty(s) {
  let match = peek(s);
  if (match === undefined) {
    return;
  }
  throw new Error(Failure, {
    cause: {
      RE_EXN_ID: Failure
    }
  });
}

function iter(f, strm) {
  let do_rec = () => {
    while (true) {
      let a = peek(strm);
      if (a === undefined) {
        return;
      }
      junk(strm);
      f(Caml_option.valFromOption(a));
      continue;
    };
  };
  do_rec();
}

function from(f) {
  return {
    count: 0,
    data: {
      TAG: "Sgen",
      _0: {
        curr: undefined,
        func: f
      }
    }
  };
}

function of_list(l) {
  return {
    count: 0,
    data: List.fold_right(((x, l) => {
      return {
        TAG: "Scons",
        _0: x,
        _1: l
      };
    }), l, "Sempty")
  };
}

function of_string(s) {
  let count = {
    contents: 0
  };
  return from(param => {
    let c = count.contents;
    if (c < s.length) {
      count.contents = count.contents + 1 | 0;
      return Caml_string.get(s, c);
    }
    
  });
}

function of_bytes(s) {
  let count = {
    contents: 0
  };
  return from(param => {
    let c = count.contents;
    if (c < s.length) {
      count.contents = count.contents + 1 | 0;
      return Caml_bytes.get(s, c);
    }
    
  });
}

function iapp(i, s) {
  return {
    count: 0,
    data: {
      TAG: "Sapp",
      _0: data(i),
      _1: data(s)
    }
  };
}

function icons(i, s) {
  return {
    count: 0,
    data: {
      TAG: "Scons",
      _0: i,
      _1: data(s)
    }
  };
}

function ising(i) {
  return {
    count: 0,
    data: {
      TAG: "Scons",
      _0: i,
      _1: "Sempty"
    }
  };
}

function lapp(f, s) {
  let f$1 = () => {
    return {
      TAG: "Sapp",
      _0: data(f()),
      _1: data(s)
    };
  };
  return {
    count: 0,
    data: {
      TAG: "Slazy",
      _0: CamlinternalLazy.from_fun(() => {
        return f$1();
      })
    }
  };
}

function lcons(f, s) {
  let f$1 = () => {
    return {
      TAG: "Scons",
      _0: f(),
      _1: data(s)
    };
  };
  return {
    count: 0,
    data: {
      TAG: "Slazy",
      _0: CamlinternalLazy.from_fun(() => {
        return f$1();
      })
    }
  };
}

function lsing(f) {
  return {
    count: 0,
    data: {
      TAG: "Slazy",
      _0: CamlinternalLazy.from_fun(() => {
        return {
          TAG: "Scons",
          _0: f(),
          _1: "Sempty"
        };
      })
    }
  };
}

function slazy(f) {
  return {
    count: 0,
    data: {
      TAG: "Slazy",
      _0: CamlinternalLazy.from_fun(() => {
        return data(f());
      })
    }
  };
}

function dump_data(f, param) {
  if (typeof param !== "object") {
    console.log("Sempty");
    return;
  }
  switch (param.TAG) {
    case "Scons" :
      console.log("Scons (");
      f(param._0);
      console.log(", ");
      dump_data(f, param._1);
      console.log(")");
      return;
    case "Sapp" :
      console.log("Sapp (");
      dump_data(f, param._0);
      console.log(", ");
      dump_data(f, param._1);
      console.log(")");
      return;
    case "Slazy" :
      console.log("Slazy");
      return;
    case "Sgen" :
      console.log("Sgen");
      return;
  }
}

function dump(f, s) {
  console.log("{count = ");
  let i = count(s);
  console.log(String(i));
  console.log("; data = ");
  dump_data(f, data(s));
  console.log("}");
  console.log("");
}

let sempty;

export {
  Failure,
  $$Error,
  from,
  of_list,
  of_string,
  of_bytes,
  iter,
  next,
  empty,
  peek,
  junk,
  count,
  npeek,
  iapp,
  icons,
  ising,
  lapp,
  lcons,
  lsing,
  sempty,
  slazy,
  dump,
}
/* No side effect */
